libs/vm/src/analyser.cpp

```diff
@@ -357,25 +390,438 @@ bool Analyser::Analyse(BlockNodePtr const &root, std::vector<std::string> &error
+  bool const objective_missing = metadata.objective_functions.empty();
+  bool const problem_missing   = metadata.problem_functions.empty();
+  bool const work_missing      = metadata.work_functions.empty();
+
+  bool const any_missing = clear_missing || objective_missing || problem_missing || work_missing;
+
+  std::ostringstream ss;
+  if (clear_missing)
+  {
+    ss << " " << CLEAR_ANNOTATION;
+  }
+  if (objective_missing)
+  {
+    ss << " " << OBJECTIVE_ANNOTATION;
+  }
+  if (problem_missing)
+  {
+    ss << " " << PROBLEM_ANNOTATION;
+  }
+  if (work_missing)
+  {
+    ss << " " << WORK_ANNOTATION;
+  }
+  std::string const missing_synergetic_names = ss.str();
+
+  auto report_missing_and_multiple_synergetics = [this, missing_synergetic_names, any_missing](
+                                                     auto const &annotation_name,
+                                                     auto const &filename_nodes) {
+    for (auto const &x : filename_nodes)
+    {
+      if (filename_nodes.size() > 1u)
+      {
+        std::ostringstream ss;
+        ss << "Multiple " << annotation_name << " functions found; only one " << annotation_name
+           << " function is allowed per contract";
+        AddError(x.filename, x.node->line, ss.str());
+      }
+
+      if (any_missing)
+      {
+        std::ostringstream ss;
+        ss << annotation_name
+           << " function found, but the following synergetic functions are missing:"
+           << missing_synergetic_names;
+        AddError(x.filename, x.node->line, ss.str());
+      }
+    }
+  };
+
+  report_missing_and_multiple_synergetics(CLEAR_ANNOTATION, metadata.clear_functions);
+  report_missing_and_multiple_synergetics(OBJECTIVE_ANNOTATION, metadata.objective_functions);
+  report_missing_and_multiple_synergetics(PROBLEM_ANNOTATION, metadata.problem_functions);
+  report_missing_and_multiple_synergetics(WORK_ANNOTATION, metadata.work_functions);
+
+  return false;
+}
+
+void Analyser::CheckSynergeticContract(LedgerRestrictionMetadata const &metadata)
+{
+  assert(metadata.clear_functions.size() == 1);
+  assert(metadata.objective_functions.size() == 1);
+  assert(metadata.problem_functions.size() == 1);
+  assert(metadata.work_functions.size() == 1);
+
+  auto const &clear_file     = metadata.clear_functions.back().filename;
+  auto const &objective_file = metadata.objective_functions.back().filename;
+  auto const &problem_file   = metadata.problem_functions.back().filename;
+  auto const &work_file      = metadata.work_functions.back().filename;
+
+  auto const &clear_node     = metadata.clear_functions.back().node;
+  auto const &objective_node = metadata.objective_functions.back().node;
+  auto const &problem_node   = metadata.problem_functions.back().node;
+  auto const &work_node      = metadata.work_functions.back().node;
+
+  auto const &clear     = *ConvertToExpressionNodePtr(clear_node->children[1])->function;
```

Comment:
Rename these clear_function etc just to make easier to read,

Reply:
done